polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -477,74 +477,110 @@ impl<Config: config::Config> XcmExecutor<Config> {
 		);
 		match instr {
 			WithdrawAsset(assets) => {
-				// Take `assets` from the origin account (on-chain) and place in holding.
-				let origin = *self.origin_ref().ok_or(XcmError::BadOrigin)?;
-				for asset in assets.into_inner().into_iter() {
-					Config::AssetTransactor::withdraw_asset(&asset, &origin, Some(&self.context))?;
-					self.subsume_asset(asset)?;
+				let old_holding = self.holding.clone();
```

Comment:
I would consider storing+restoring holding in all instructions as long as they call into `TransactionalProcessor::process` in an automatic way that is more error-prone.

Reply:
That's actually what Gav advised against in his comments above -- some instructions don't need transactionality.